src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,239 @@
+| add_compat_left {a b c : pre R M} : rel a b → rel (a + c) (b + c)
+| add_compat_right {a b c : pre R M} : rel a b → rel (c + a) (c + b)
+| mul_compat_left {a b c : pre R M} : rel a b → rel (a * c) (b * c)
+| mul_compat_right {a b c : pre R M} : rel a b → rel (c * a) (c * b)
+
+end tensor_algebra
+
+/--
+The tensor algebra of the module `M` over the commutative semiring `R`.
+-/
+def tensor_algebra := quot (tensor_algebra.rel R M)
+
+namespace tensor_algebra
+
+local attribute [instance]
+  pre.has_coe_module pre.has_coe_semiring pre.has_mul pre.has_add pre.has_zero
+  pre.has_one pre.has_scalar
+
+instance : semiring (tensor_algebra R M) :=
+{ add := quot.map₂ (+) (λ _ _ _, rel.add_compat_right) (λ _ _ _, rel.add_compat_left),
+  add_assoc := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.add_assoc },
+  zero := quot.mk _ 0,
+  zero_add := by { rintro ⟨⟩, exact quot.sound rel.zero_add },
+  add_zero := begin
+    rintros ⟨⟩,
+    change quot.mk _ _ = _,
+    rw [quot.sound rel.add_comm, quot.sound rel.zero_add],
+  end,
+  add_comm := by { rintros ⟨⟩ ⟨⟩, exact quot.sound rel.add_comm },
+  mul := quot.map₂ (*) (λ _ _ _, rel.mul_compat_right) (λ _ _ _, rel.mul_compat_left),
+  mul_assoc := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.mul_assoc },
+  one := quot.mk _ 1,
+  one_mul := by { rintros ⟨⟩, exact quot.sound rel.one_mul },
+  mul_one := by { rintros ⟨⟩, exact quot.sound rel.mul_one },
+  left_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.left_distrib },
+  right_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.right_distrib },
+  zero_mul := by { rintros ⟨⟩, exact quot.sound rel.zero_mul },
+  mul_zero := by { rintros ⟨⟩, exact quot.sound rel.mul_zero } }
+
+instance : inhabited (tensor_algebra R M) := ⟨0⟩
+
+instance : has_scalar R (tensor_algebra R M) :=
+{ smul := λ r a, quot.lift_on a (λ x, quot.mk _ $ ↑r * x) $
+  λ a b h, quot.sound (rel.mul_compat_right h) }
+
+instance : algebra R (tensor_algebra R M) :=
+{ to_fun := λ r, quot.mk _ r,
+  map_one' := rfl,
+  map_mul' := λ _ _, quot.sound rel.mul_scalar,
+  map_zero' := rfl,
+  map_add' := λ _ _, quot.sound rel.add_scalar,
+  commutes' := λ _, by { rintros ⟨⟩, exact quot.sound rel.central_scalar },
+  smul_def' := λ _ _, rfl }
+
+/--
+The canonical linear map `M →ₗ[R] tensor_algebra R M`.
+-/
+def ι : M →ₗ[R] (tensor_algebra R M) :=
+{ to_fun := λ m, quot.mk _ m,
+  map_add' := λ x y, quot.sound rel.add_lin,
+  map_smul' := λ r x, quot.sound rel.smul_lin }
+
+/--
+Given a linear map `f : M → A` where `A` is an `R`-algebra,
+`lift R M f` is the unique lift of `f` to a morphism of `R`-algebras
+`tensor_algebra R M → A`.
+-/
+def lift {A : Type u₃} [semiring A] [algebra R A] (f : M →ₗ[R] A) : tensor_algebra R M →ₐ[R] A :=
+{ to_fun := λ a, quot.lift_on a (lift_fun _ _ f) $ λ a b h,
+  begin
+    induction h,
+    { change f _ = f _ + f _, simp, },
+    { change f _ = (algebra_map _ _ _) * f _, rw linear_map.map_smul,
+      exact algebra.smul_def h_r (f h_a) },
```

Comment:
```suggestion
    { change f _ = (algebra_map _ _ _) * f _, 
      rw linear_map.map_smul,
      exact algebra.smul_def h_r (f h_a) },
```
etc. (Although I admit I'm sloppy about this; if it really feels like two tactics are doing one thing, I guess on a single line is fine.)